packages/frontend-2/lib/viewer/composables/filtering/filtering.ts

```diff
@@ -0,0 +1,793 @@
+import type {
+  PropertyInfo,
+  NumericPropertyInfo,
+  StringPropertyInfo
+} from '@speckle/viewer'
+
+import { FilteringExtension } from '@speckle/viewer'
+import { difference, uniq, partition } from 'lodash-es'
+import {
+  useInjectedViewerState,
+  type InjectableViewerState
+} from '~~/lib/viewer/composables/setup'
+import { isNumericPropertyInfo } from '~/lib/viewer/helpers/sceneExplorer'
+import {
+  type FilterCondition,
+  FilterLogic,
+  FilterType,
+  type FilterData,
+  type NumericFilterData,
+  type StringFilterData,
+  type CreateFilterParams,
+  isNumericFilter,
+  NumericFilterCondition,
+  StringFilterCondition,
+  ExistenceFilterCondition,
+  SortMode,
+  type DataSlice,
+  type QueryCriteria,
+  type ValueGroupsMap,
+  type ValueGroupMapItem
+} from '~/lib/viewer/helpers/filters/types'
+import { getConditionLabel } from '~/lib/viewer/helpers/filters/constants'
+import { useOnViewerLoadComplete } from '~~/lib/viewer/composables/viewer'
+import { useFilteringDataStore } from '~/lib/viewer/composables/filtering/dataStore'
+import {
+  shouldExcludeFromFiltering,
+  getPropertyName,
+  isKvpFilterable,
+  getFilterDisabledReason,
+  findFilterByKvp
+} from '~/lib/viewer/helpers/filters/utils'
+import { useFilterColors } from '~/lib/viewer/composables/filtering/coloring'
+
+export function useFilterUtilities(
+  options?: Partial<{ state: InjectableViewerState }>
+) {
+  const state = options?.state || useInjectedViewerState()
+  const {
+    viewer,
+    ui: { filters, explodeFactor }
+  } = state
+
+  const dataStore = useFilteringDataStore()
+  const { removeColorFilter } = useFilterColors({ state })
+
+  const { instance } = viewer
+  const { resourceItems } = state.resources.response
+
+  const populateInternalDataStore = async () => {
+    const tree = instance.getWorldTree()
+    if (!tree || !resourceItems.value.length) {
+      return
+    }
+
+    const availableResources = resourceItems.value.filter((item) => {
+      const nodes = tree.findId(item.objectId)
+      return nodes && nodes.length > 0
+    })
+
+    if (availableResources.length === 0) {
+      return
+    }
+
+    dataStore.clearDataOnRouteLeave()
+
+    const resources = availableResources.map((item) => ({
+      resourceUrl: item.objectId
+    }))
+
+    await dataStore.populateDataStore(instance, resources)
+  }
+
+  useOnViewerLoadComplete(async () => {
+    await populateInternalDataStore()
+  })
+
+  // Watch for resource changes (e.g., when switching versions) and repopulate the data store
+  watch(
```

Comment:
another composable where utils are mixed with core/low-level integration work. the implication behind the name is that you can import this multiple times from different components (and we do this), but then there's also integration/setup work in here that should only be ever invoked once on setup and yet gets invoked every single time this composable is invoked

so for like the 50 or so components that run this composable, all of them have this watcher that populate the internal data store 50 times